libs/ktem/ktem/reasoning/react.py

```diff
@@ -0,0 +1,322 @@
+                        + retrieved_content
+                        + " \n<br>"
+                    )
+
+            print("Retrieved #{}: {}".format(_id, retrieved_content[:100]))
+            print("Score", retrieved_item.metadata.get("relevance_score", None))
+
+        # trim context by trim_len
+        if evidence:
+            text_splitter = CharacterTextSplitter.from_tiktoken_encoder(
+                chunk_size=trim_len,
+                chunk_overlap=0,
+                separator=" ",
+                model_name="gpt-3.5-turbo",
+            )
+        texts = text_splitter.split_text(evidence)
+        evidence = texts[0]
+
+        return Document(content=evidence)
+
+
+TOOL_REGISTRY = {
+    "Google": GoogleSearchTool(),
+    "Wikipedia": WikipediaTool(),
+    "LLM": LLMTool(),
+    "SearchDoc": DocSearchTool(),
+}
+
+DEFAULT_QA_PROMPT = (
+    "Answer the following questions as best you can. Give answer in {lang}. "
+    "You have access to the following tools:\n"
+    "{tool_description}\n"
+    "Use the following format:\n\n"
+    "Question: the input question you must answer\n"
+    "Thought: you should always think about what to do\n\n"
+    "Action: the action to take, should be one of [{tool_names}]\n\n"
+    "Action Input: the input to the action, should be different from the action input "
+    "of the same action in previous steps.\n\n"
+    "Observation: the result of the action\n\n"
+    "... (this Thought/Action/Action Input/Observation can repeat N times)\n"
+    "#Thought: I now know the final answer\n"
+    "Final Answer: the final answer to the original input question\n\n"
+    "Begin! After each Action Input.\n\n"
+    "Question: {instruction}\n"
+    "Thought: {agent_scratchpad}\n"
+)
+
+DEFAULT_REWRITE_PROMPT = (
+    "Given the following question, rephrase and expand it "
+    "to help you do better answering. Maintain all information "
+    "in the original question. Keep the question as concise as possible. "
+    "Give answer in {lang}\n"
+    "Original question: {question}\n"
+    "Rephrased question: "
+)
+
+
+class RewriteQuestionPipeline(BaseComponent):
+    """Rewrite user question
+
+    Args:
+        llm: the language model to rewrite question
+        rewrite_template: the prompt template for llm to paraphrase a text input
+        lang: the language of the answer. Currently support English and Japanese
+    """
+
+    llm: ChatLLM = Node(default_callback=lambda _: llms.get_default())
+    rewrite_template: str = DEFAULT_REWRITE_PROMPT
+
+    lang: str = "English"
+
+    def run(self, question: str) -> Document:  # type: ignore
+        prompt_template = PromptTemplate(self.rewrite_template)
+        prompt = prompt_template.populate(question=question, lang=self.lang)
+        messages = [
+            SystemMessage(content="You are a helpful assistant"),
+            HumanMessage(content=prompt),
+        ]
+        return self.llm(messages)
+
+
+class ReactAgentPipeline(BaseReasoning):
```

Comment:
should take tools as input param instead of fixed TOOL_REGISTRY.

Reply:
https://github.com/Cinnamon/kotaemon/issues/63